akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala

```diff
@@ -131,93 +131,94 @@ private[akka] final class ReplayingEvents[C, E, S](
-                    seenSeqNrPerReplica = state.seenSeqNrPerReplica + (meta.originReplica -> meta.originSequenceNr))
-                  replication.clearContext()
-                case Some((_, _, replication)) =>
-                  replication.clearContext()
-                  state = state.copy(state = newState, eventSeenInInterval = true)
-                case _ =>
-                  state = state.copy(state = newState, eventSeenInInterval = true)
-              }
+            val newState = setup.eventHandler(state.state, event)
+
+            setup.replication match {
+              case Some(replication) =>
+                replication.clearContext()
+              case None =>
             }
 
-            eventSeq match {
-              case SingleEventSeq(event) => handleEvent(event)
-              case EventsSeq(events)     => events.foreach(handleEvent)
-              case EmptyEventSeq         => // no events
+            replicatedMetaAndSelfReplica match {
+              case Some((meta, selfReplica, replication)) if meta.originReplica != selfReplica =>
+                // keep track of highest origin seqnr per other replica
+                state = state.copy(
+                  state = newState,
+                  eventSeenInInterval = true,
+                  version = meta.version,
+                  seenSeqNrPerReplica = state.seenSeqNrPerReplica + (meta.originReplica -> meta.originSequenceNr))
+                replication.clearContext()
+              case Some((_, _, replication)) =>
+                replication.clearContext()
+                state = state.copy(state = newState, eventSeenInInterval = true)
+              case _ =>
+                state = state.copy(state = newState, eventSeenInInterval = true)
             }
+          }
 
-            this
-          } catch {
-            case NonFatal(ex) =>
-              state = state.copy(repr.sequenceNr)
-              onRecoveryFailure(ex, eventForErrorReporting.toOption)
+          eventSeq match {
+            case SingleEventSeq(event) => handleEvent(event)
+            case EventsSeq(events)     => events.foreach(handleEvent)
+            case EmptyEventSeq         => // no events
           }
 
-        case RecoverySuccess(highestJournalSeqNr) =>
+          this
+        } catch {
+          case NonFatal(ex) =>
+            state = state.copy(repr.sequenceNr)
+            onRecoveryFailure(ex, eventForErrorReporting.toOption, "replaying-event")
+        }
+
+      case RecoverySuccess(highestJournalSeqNr) =>
+        try {
           val highestSeqNr = Math.max(highestJournalSeqNr, state.seqNr)
           state = state.copy(seqNr = highestSeqNr)
           setup.internalLogger.debug("Recovery successful, recovered until sequenceNr: [{}]", highestSeqNr)
           onRecoveryCompleted(state)
+        } catch {
+          case ex: UnstashException[_] =>
+            // let supervisor handle it, don't treat as recovery failure
+            throw ex
```

Comment:
I think that it is correct that unstash exceptions can only ever happen in onRecoveryComplete.